src/hotspot/share/services/nmtPreInit.hpp

```diff
@@ -0,0 +1,353 @@
+//
+// The problem is that inside os::free(p), we only get an opaque void* p; we do not know if p had been
+// allocated in (a) or (b) phase. Therefore, we do not know if p is preceded by an NMT header which we
+// would need to subtract from the pointer before calling free(3). There is no safe way to "guess" here
+// without risking C-heap corruption.
+//
+// To solve this, we need a way to quickly determine, at os::free(p), whether p was a pre-NMT-init
+// allocation. There are several ways to do this, see discussion under JDK-8256844.
+//
+// One of the easiest and most elegant ways is to store early allocation pointers in a lookup table.
+// This is what NMTPreInit does.
+//
+//////////////////////////////////////////////////////////////////////////
+//
+// VM initialization wrt NMT:
+//
+//---------------------------------------------------------------
+//-> launcher dlopen's libjvm                           ^
+//   -> dynamic C++ initialization                      |
+//           of libjvm                                  |
+//                                                      |
+//-> launcher starts new thread (maybe)          NMT pre-init phase : store allocated pointers in lookup table
+//                                                      |
+//-> launcher invokes CreateJavaVM                      |
+//   -> VM initialization before arg parsing            |
+//   -> VM argument parsing                             v
+//   -> NMT initialization  -------------------------------------
+//                                                      ^
+//   ...                                                |
+//   -> VM life...                               NMT post-init phase : lookup table is read-only; use it in os::free() and os::realloc().
+//   ...                                                |
+//                                                      v
+//----------------------------------------------------------------
+//
+//////////////////////////////////////////////////////////////////////////
+//
+// Notes:
+// - The VM will malloc() and realloc() several thousand times before NMT initialization.
+//   Starting with a lot of arguments increases this number since argument parsing strdups
+//   around a lot.
+// - However, *surviving* allocations (allocations not freed immediately) are much rarer:
+//   typically only about 300-500. Again, mainly depending on the number of VM args.
+// - There are a few cases of pre-to-post-init reallocs where pre-init allocations get
+//   reallocated after NMT initialization. Those we need to handle with special care (see
+//   NMTPreInit::handle_realloc()). Because of them we need to store allocation size with
+//   every pre-init allocation.
+
+// For the lookup table, design considerations are:
+//   - lookup speed is paramount since lookup is done for every os::free() call.
+//   - insert/delete speed only matters for VM startup - after NMT initialization the lookup
+//     table is readonly
```

Comment:
The point is *after NMT initialization* - `find_and_remove` only gets called before NMT initialization; after that, we only do non-modifying lookup. You'll find the logic in `NMTPreInit::handle_realloc()` and `NMTPreInit::handle_free()`, respectively.

The basic idea behind this is that we remove pointers from the map as long as we can without risking concurrency issues, which is until NMT initialization. After that, we leave the map alone. It was either that or protect the map with a lock, and this is the lesser of two evils since the map is usually sparsely populated.